.github/workflows/cd.yml

```diff
@@ -6,54 +8,71 @@ on:
     branches:
       - main
 
+  workflow_dispatch:
+    inputs:
+      ref:
+        description: "Branch or commit SHA to tag"
+        type: string
+        required: true
+      tag:
+        description: "Tag for release"
+        type: string
+        required: true
+
 jobs:
-  publish:
-    # This job will only run if the PR has been merged (and not closed without merging).
-    if: "github.event.pull_request.merged == true && !contains(github.event.pull_request.head.message, 'skipci')"
+  release:
+    if: "${{ github.event_name == 'workflow_dispatch' || github.event.pull_request.merged == true }}"
     runs-on: ubuntu-latest
     steps:
-      - name: Checkout repository
-        uses: actions/checkout@v3
-
-      - name: Test package is publishable with PyPI test server
-        uses: JRubics/poetry-publish@v1.13
+      - name: Checkout Repository
+        if: ${{ github.event_name != 'workflow_dispatch' }}
+        uses: actions/checkout@v4
         with:
-          python_version: '3.9'
-          pypi_token: ${{ secrets.TEST_PYPI_TOKEN }}
-          repository_name: 'testpypi'
-          repository_url: 'https://test.pypi.org/legacy/'
+          ref: ${{ inputs.ref }}
 
-      - name: Publish latest package to PyPI
-        uses: JRubics/poetry-publish@v1.16
-        with:
-          python_version: '3.9'
-          poetry_version: '==1.4.2' # (PIP version specifier syntax)
-          pypi_token: ${{ secrets.PYPI_TOKEN }}
-          ignore_dev_requirements: 'yes'
-
-  release:
-    # This job will only run if the PR has been merged (and not closed without merging).
-    if: "github.event.pull_request.merged == true && !contains(github.event.pull_request.head.message, 'skipci')"
-    runs-on: ubuntu-latest
-    needs: [publish]
-    steps:
-      - name: Checkout code
-        uses: actions/checkout@v3
+      - name: Install Poetry
+        if: ${{ github.event_name != 'workflow_dispatch' }}
+        uses: snok/install-poetry@v1.3.2
 
-      - name: Install poetry
-        run: pip install poetry
+      - name: Check pyproject.toml file
+        if: ${{ github.event_name != 'workflow_dispatch' }}
+        run: poetry check
 
       - name: Get package version
-        id: version
-        run: echo "::set-output name=package_version::$(poetry version -s)"
+        id: get-package-version
+        if: ${{ github.event_name != 'workflow_dispatch' }}
+        run: echo "package_version=$(poetry version -s)" >> $GITHUB_OUTPUT
 
-      - name: Create release
+      - name: Create Release
         uses: actions/create-release@v1
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # This token is provided by Actions, no need to create your own.
         with:
-          tag_name: ${{ steps.version.outputs.package_version }}
+          commitish: ${{ inputs.ref }}
+          tag_name: ${{ inputs.tag || steps.get-package-version.outputs.package_version }}
```

Comment:
Again, this allows you to make a tag totally different than the package. There should be a single source of truth about the version; the pyproject.toml file.